Text-Input State

bbqbaron opened the pull request:
A sketch for managing the state of the text input, to avoid blowing away user input during a re-render. From  #24. The objective is to make the picker retain an idea of its text input's state, and not overwrite whatever the user happens to be typing in the case of a DOM reflow, time-based model change, or any other incidental event.

The changes, in natural language so we know they're comprehensible. I suspect that there might be an _even more_ parsimonious minimal way to express absolutely correct state at all times, but I don't see any holes in this one...yet.

- add new state vars `editing` and `textInput`
- `editing` toggles as the text input gains and loses focus.
- `inputText` is probably best expressed as a merge of the following changes:
  - picking a date with the picker sets it to the formatted date
  - entering or leaving `editing` sets if to the formatted picked date, if any
  - typing in the field updates it with the raw typed value
- if `editing`, render the text input with the raw text input value. otherwise, render it with the formatted date. this is redundant with the second source of changes above, since presumably outside of `editing`, the input text _is_ the formatted date anyway, but I'm not infallible!

bbqbaron commented:
You could probably remove `editing`, just make sure you got all the synchronization of `inputText` right, and always render it as-is.

A5308Y commented:
Thanks @bbqbaron! It works perfectly for my case! I agree you could probably get rid of editing somehow, but from my external perspective everything's perfect :-)

bbqbaron commented:
@A5308Y I realized that I'd also changed the behavior of the picker, since it was now changing date values with each `input` event, whereas before it was on `change`. I.e. it used to be like a tiny form and now it was super-eager. I'm not sure I feel strongly about either behavior, but I didn't want to break anyone's assumptions about how `update` works without a good reason. Perhaps they had a validator with error messages running, and would now be spammed by date "changes" that the user never really intended.

So, I moved the actual date change to `Blur`. Based on my understanding of your use case, that's still correct, right? You type in field, your input still isn't erased, and you tab out to commit your change?

(Also removed `editing` while I was at it, since it was increasingly superfluous.)

Thanks for collaborating on this!

A5308Y commented:
@bbqbaron I like this version quite a bit less as it does not register the change on hitting "enter". It's a bit confusing to enter (or change) the date, hit enter and everything gets discarded.

I also found it reassuring that I could see the effect of date changes in the picker directly. I knew the change was understood and could even validate the new day, checking, for example, what weekday my entered date was.

So I have a strong preference for the first version. I hope the two arguments are a good enough reason to break assumptions about how update works.

If not: Maybe it could be a setting to configure the datepicker to be "eager"?

A5308Y commented:
Oh and 👍  for getting rid of `editing` :-)

btw: What does "WTB" mean? The urban dictionary didn't help me here...

bbqbaron commented:
Having mulled some more, I've taken kind of an about-face and changed as little as I could to address the particular problem in the issue (i.e., the lack of internal text state causing render bugs). I think you're right about live feedback being super helpful; I just worry that it's basically orthogonal.

When it comes to avoiding scope creep, I like to ask "How much harder is the current problem to solve if I defer the extra scope?" I think the answer here is "Not very," since presumably live updates was just a fortuitous side effect of the original set of changes I made. I'd like to open a ticket requesting that behavior, and we can take the time to make the "date did change" API more robust/configurable. +/-?

(Certainly let me know if this changeset still fixes the time-based rendering bug!)

bbqbaron commented:
Oh, also, I think the failure of Enter to update the date was because I'd attached the date change to `Blur`. Now that it's back on `Change` it seems to take.

A5308Y commented:
@bbqbaron Thanks for your reply and your work! I'd love to have the instant feedback as soon as possible (now that I know how it could be I'm hooked ;-) ), but I do understand your concerns.
I can confirm that the text rendering bug is fixed in this version and submitting on enter works as well.

But it seems like `pickedDate` isn't taken into account anymore in this version. I might've made a mistake in setting it up in my app, but with all other versions I set a preset date for the datepicker-input successfully. In this version it always starts out blank.
I've also just copied the datepicker code in elm-stuff/packages... so that might be weird as well.

But maybe you could have a quick look and see if `pickedDate` being ignored in this version seems plausible.

A5308Y commented:
On a side note: I've successfully disabled all past dates for the datepicker in my app, but I had to pass the current date to my `isDisabled` function. I realized that the datepicker must have a notion of "today", because it marks today's date and it feels a bit weird to pass the current time to the picker even though it already knows about it.

I'm okay with the way it works now, but it might be a common use case to disable past dates (making appointments, due dates, ... only make sense for the future) and it might make the datepicker easier to configure, if the datepicker's time was accessible in `isDisabled` or you could disabled past dates in the settings directly.

On the other hand it might also be confusing for users with other use cases... Well it's just a thought. Maybe it'll spark some more on your end :-)

bbqbaron commented:
`pickedDate` was definitely broken. I removed the superfluous "always set text on focus", which was masking the fact that `inputText` was being initialized unconditionally to `""`, when of course it should be the formatted picked date if any! That is now fixed, although it makes me wish the data model were more properly declarative. This doesn't have a proper SSOT arrangement, clearly.

Instant feedback should be a thing! I'll start by opening an issue for it, then work can start on it separately. I suspect it'll be possible to rig up "You get instant visual feedback but the change doesn't persist right away"; it'll just mean having a couple of dates on the model and being smart about handling them.

Could you open an issue for the `isDisabled` behavior? I'd hate for it to go unnoticed in the depths of a PR. Just a description of current and desired behavior, etc.

bbqbaron commented:
Opened #31.
